### PR TITLE
[#1917] fix(server,netty): Fix memory leak issues when handling SendShuffleDataRequest

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes were proposed in this pull request?

Extract a method to release resources including Netty buffers and memory metrics when handling SendShuffleDataRequest.

### Why are the changes needed?

Fix: #1917

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Tested on our test cluster

### CODE REVIEW

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java [hunk 1]
```diff
@@ -130,6 +165,17 @@ public void handleSendShuffleDataRequest(TransportClient client, SendShuffleData
       rpcResponse =
           new RpcResponse(req.getRequestId(), StatusCode.STAGE_RETRY_IGNORE, responseMessage);
       client.getChannel().writeAndFlush(rpcResponse);
+      // TODO(baoloongmao): release memory and buffer.
```
Comment (rickyma) [inline server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:168]: We should fix this as well.

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java [hunk 2]
```diff
@@ -309,6 +292,37 @@ public void handleSendShuffleDataRequest(TransportClient client, SendShuffleData
...
+          .flatMap(Collection::stream)
+          .forEach(
+              block -> {
+                block.getData().release();
+                counter.getAndAdd(block.getData().readableBytes());
+              });
+      ShuffleServerMetrics.counterTotalReceivedDataSize.inc(counter.get());
+    } else {
+      LOG.error(
+          "Failed to handle send shuffle data request, no blocks this request. appId: {}, shuffleId: {}, requireBufferId: {}",
```
Comment (rickyma) [inline server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:319]: no blocks found in this request

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java [hunk 3]
```diff
@@ -309,6 +292,37 @@ public void handleSendShuffleDataRequest(TransportClient client, SendShuffleData
...
+    }
+    if (MapUtils.isNotEmpty(req.getPartitionToBlocks())) {
+      AtomicLong counter = new AtomicLong(0);
+      // release memory
+      req.getPartitionToBlocks().values().stream()
+          .flatMap(Collection::stream)
+          .forEach(
+              block -> {
+                block.getData().release();
+                counter.getAndAdd(block.getData().readableBytes());
```
Comment (rickyma) [inline server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:314]: remove counter

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java [hunk 4]
```diff
@@ -309,6 +292,37 @@ public void handleSendShuffleDataRequest(TransportClient client, SendShuffleData
     client.getChannel().writeAndFlush(rpcResponse);
   }
 
+  private static void releaseBufferMemoryAndMetrics(
```
Comment (rickyma) [inline server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:295]: releaseNettyBufferAndMetrics

**Review by zuston — COMMENTED** [review]
Overall LGTM. Could you help use the unit test to cover this case? @maobaolong

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java (hunk 1 above)
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java]: done

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java (hunk 2 above)
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:319]: done

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java (hunk 3 above)
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:314]: done

File: server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java (hunk 4 above)
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java:295]: done

**Comment by maobaolong** [comment]
@rickyma All the comments have been addressed, PTAL

**Review by rickyma — APPROVED** [review]
LGTM

**Comment by rickyma** [comment]
Merged to master.
Thanks for your contribution @maobaolong  
Thanks for your review @zuston

**Comment by maobaolong** [comment]
> Overall LGTM. Could you help use the unit test to cover this case? @maobaolong

Thank for your suggestion, as discussed with @zuston in wechat, i will introduce a ut by another PR.

**Pull request merged**